File: backend/app/routers/toilets.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
import redis.asyncio as aioredis
from typing import Optional

from app.dependencies import get_db, get_redis, get_current_user, get_optional_user, require_role
from app.models.user import User
from app.models.enums import UserRole, Gender, ToiletType, PaperType
from app.schemas.toilet import ToiletCreate, ToiletUpdate, NearbyResponse, ToiletDetail
from app.services.toilet import (
    create_toilet, get_toilet, update_toilet,
    soft_delete_toilet, hard_delete_toilet, nearby, search,
    verify_toilet, confirm_toilet,
)

router = APIRouter(prefix="/api/v1/toilets", tags=["toilets"])


@router.get("/nearby", response_model=NearbyResponse)
async def nearby_endpoint(
    lat: float = Query(..., ge=-90, le=90),
    lon: float = Query(..., ge=-180, le=180),
    radius: float = Query(1000, le=10000),
    gender: Optional[Gender] = None,
    is_free: Optional[bool] = None,
    has_water: Optional[bool] = None,
    has_hot_water: Optional[bool] = None,
    has_soap: Optional[bool] = None,
    has_dryer: Optional[bool] = None,
    paper_type: Optional[PaperType] = None,
    has_child_toilet: Optional[bool] = None,
    has_changing_table: Optional[bool] = None,
    is_accessible: Optional[bool] = None,
    toilet_type: Optional[ToiletType] = None,
    is_open_now: Optional[bool] = None,
    min_rating: Optional[float] = None,
    limit: int = Query(20, ge=1, le=50),
    cursor: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    redis: aioredis.Redis = Depends(get_redis),
):
    filters = {
        "gender": gender,
        "is_free": is_free,
        "has_water": has_water,
        "has_hot_water": has_hot_water,
        "has_soap": has_soap,
        "has_dryer": has_dryer,
        "paper_type": paper_type,
        "has_child_toilet": has_child_toilet,
        "has_changing_table": has_changing_table,
        "is_accessible": is_accessible,
        "toilet_type": toilet_type,
        "is_open_now": is_open_now,
        "min_rating": min_rating,
    }
    return await nearby(db, redis, lat, lon, radius, limit, cursor, **filters)


@router.get("/search", response_model=NearbyResponse)
async def search_endpoint(
    q: str = Query(..., min_length=3),
    lat: Optional[float] = None,
    lon: Optional[float] = None,
    limit: int = Query(20, ge=1, le=50),
    cursor: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
):
    return await search(db, q, lat, lon, limit, cursor)


@router.get("/{toilet_id}")
async def get_toilet_endpoint(
    toilet_id: str,
    db: AsyncSession = Depends(get_db),
    redis: aioredis.Redis = Depends(get_redis),
    user: Optional[User] = Depends(get_optional_user),
):
    return await get_toilet(db, redis, toilet_id, user)


@router.post("/", status_code=201)
async def create_toilet_endpoint(
    data: ToiletCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    toilet = await create_toilet(db, data.model_dump(exclude_unset=True), str(user.id))
    return {"id": str(toilet.id), "name": toilet.name}


@router.patch("/{toilet_id}")
async def update_toilet_endpoint(
    toilet_id: str,
    data: ToiletUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    toilet = await update_toilet(db, toilet_id, data.model_dump(exclude_unset=True, exclude_none=True), user)
    return {"id": str(toilet.id), "name": toilet.name}


@router.delete("/{toilet_id}", status_code=204)
async def delete_toilet_endpoint(
    toilet_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await soft_delete_toilet(db, toilet_id, user)


@router.delete("/{toilet_id}/hard", status_code=204)
async def hard_delete_toilet_endpoint(
    toilet_id: str,
    user: User = Depends(require_role(UserRole.MODERATOR, UserRole.ADMIN)),
    db: AsyncSession = Depends(get_db),
):
    await hard_delete_toilet(db, toilet_id)


@router.post("/{toilet_id}/confirm", status_code=201)
async def confirm_toilet_endpoint(
    toilet_id: str,
    is_actual: bool = True,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    confirmation = await confirm_toilet(db, toilet_id, str(user.id), is_actual)
    return {"id": str(confirmation.id)}


@router.patch("/{toilet_id}/verify")
async def verify_toilet_endpoint(
    toilet_id: str,
    user: User = Depends(require_role(UserRole.MODERATOR, UserRole.ADMIN)),
    db: AsyncSession = Depends(get_db),
):
    toilet = await verify_toilet(db, toilet_id)
    return {"id": str(toilet.id), "is_verified": toilet.is_verified}

File: frontend/src/api/toilets.ts
import { apiClient } from './client'
import type { Toilet, ToiletCreate } from '@/types/toilet'
import type { PaginatedResponse, NearbyParams, SearchParams } from '@/types/api'

export async function getNearbyToilets(params: NearbyParams): Promise<PaginatedResponse<Toilet>> {
  const searchParams: Record<string, string> = {
    lat: String(params.lat),
    lon: String(params.lon),
  }
  if (params.radius) searchParams.radius = String(params.radius)
  if (params.gender) searchParams.gender = params.gender
  if (params.is_free !== undefined) searchParams.is_free = String(params.is_free)
  if (params.has_water !== undefined) searchParams.has_water = String(params.has_water)
  if (params.has_hot_water !== undefined) searchParams.has_hot_water = String(params.has_hot_water)
  if (params.has_soap !== undefined) searchParams.has_soap = String(params.has_soap)
  if (params.has_dryer !== undefined) searchParams.has_dryer = String(params.has_dryer)
  if (params.is_accessible !== undefined) searchParams.is_accessible = String(params.is_accessible)
  if (params.has_child_toilet !== undefined) searchParams.has_child_toilet = String(params.has_child_toilet)
  if (params.has_changing_table !== undefined) searchParams.has_changing_table = String(params.has_changing_table)
  if (params.paper_type) searchParams.paper_type = params.paper_type
  if (params.is_open_now !== undefined) searchParams.is_open_now = String(params.is_open_now)
  if (params.min_rating !== undefined) searchParams.min_rating = String(params.min_rating)
  if (params.toilet_type) searchParams.toilet_type = params.toilet_type
  if (params.cursor) searchParams.cursor = params.cursor
  if (params.limit) searchParams.limit = String(params.limit)

  return apiClient.get('toilets/nearby', { searchParams }).json<PaginatedResponse<Toilet>>()
}

export async function searchToilets(params: SearchParams): Promise<PaginatedResponse<Toilet>> {
  const searchParams: Record<string, string> = { q: params.q }
  if (params.lat !== undefined) searchParams.lat = String(params.lat)
  if (params.lon !== undefined) searchParams.lon = String(params.lon)
  if (params.cursor) searchParams.cursor = params.cursor
  if (params.limit) searchParams.limit = String(params.limit)
  return apiClient.get('toilets/search', { searchParams }).json<PaginatedResponse<Toilet>>()
}

export async function getToilet(id: string): Promise<Toilet> {
  return apiClient.get(`toilets/${id}`).json<Toilet>()
}

export async function createToilet(data: ToiletCreate): Promise<Toilet> {
  return apiClient.post('toilets', { json: data }).json<Toilet>()
}

export async function updateToilet(id: string, data: Partial<ToiletCreate>): Promise<Toilet> {
  return apiClient.patch(`toilets/${id}`, { json: data }).json<Toilet>()
}

export async function deleteToilet(id: string): Promise<void> {
  await apiClient.delete(`toilets/${id}`)
}

export async function uploadToiletPhoto(toiletId: string, file: File): Promise<{ url: string }> {
  const formData = new FormData()
  formData.append('file', file)
  return apiClient.post(`toilets/${toiletId}/photos`, { body: formData }).json<{ url: string }>()
}

export async function getFavorites(cursor?: string, limit?: number): Promise<PaginatedResponse<Toilet>> {
  const searchParams: Record<string, string> = {}
  if (cursor) searchParams.cursor = cursor
  if (limit) searchParams.limit = String(limit)
  return apiClient.get('favorites', { searchParams }).json<PaginatedResponse<Toilet>>()
}

export async function addToFavorite(toiletId: string): Promise<void> {
  await apiClient.post(`toilets/${toiletId}/favorite`)
}

export async function removeFromFavorite(toiletId: string): Promise<void> {
  await apiClient.delete(`toilets/${toiletId}/favorite`)
}
